packages/pkl.experimental.structuredRead/structuredRead.pkl

```diff
@@ -0,0 +1,261 @@
+      acc + findFailures(value)
+    else
+      acc
+  )
+
+abstract class Failure {
+  message: String
+  path: List<String>
+}
+
+class ReadFailure extends Failure
+class CoerceFailure extends Failure
+
+local function ReadFail(path: List<String>, msg: String) = let (_path = path ) new ReadFailure {
+  message = msg
+  path = _path
+}
+
+local function CoerceFail(path: List<String>, msg: String) = let (_path = path ) new CoerceFailure {
+  message = msg
+  path = _path
+}
+
+local function parseArraySyntax(input: String): Listing = new {
+  when (input.matches(Regex(#"^\{.*}$"#))) {
+    ...input.drop(1).dropLast(1).split(",")
+  } else {
+    input
+  }
+}
+
+local function applyToDynamic(path, default: Dynamic) =
+  if (default.length() == 0 && default.toMap().length == 0)
+    // Do some special handling for empty dynamic types
+     let (potentialValue = readValue(path))
+       if (potentialValue is ReadFailure)
+         default
+       else if (potentialValue != null)
+         if (potentialValue is String)
+           parseArraySyntax(potentialValue)
+         else
+           potentialValue
+       else
+         default
+  else
+    new Mapping {
+      for (name, defaultValue in default) {
+        when (defaultValue is Dynamic) {
+          [name] = applyToDynamic(path.add(name.toString()), defaultValue)
+        } else {
+          [name] = let (_path = path.add(name.toString()))
+            let (potentialValue = coerceInput(readValue(_path), defaultValue.getClass(), _path))
+            if (potentialValue is ReadFailure)
+              defaultValue
+            else
+              potentialValue
+        }
+      }
+    }.toMap().toDynamic()
+
+hidden classHandlers: Mapping<Class, (String) -> Any> = new {
+  [Int] = (value) ->
+    value.toIntOrNull()
+
+  [String] = (value) ->
+    value
+
+  [Float] = (value) ->
+    value.toFloatOrNull()
+
+  [Number] = (value) ->
+    value.toIntOrNull() ?? value.toFloatOrNull()
+
+  [Boolean] = (value) ->
+    value.toBooleanOrNull()
+
+  [Null] = (_) -> throw("""
+    Unable to perform structured read due to ambiguous type in schema.
+    This is normally caused by a schema property having default value of `null` but no explicit type
+  """)
+
+  [Resource] = (value) -> value
+}
+
+local function coerceInput(value: String?|Resource?, clazz: Class|TypeAlias, path:List<String>) =
+    if (classHandlers.containsKey(clazz))
+      if (value == null)
+        ReadFail(
+          path,
+          "Not provided. Make sure you pass the argument '-p \(path.join(pathSeperator))=<\(clazz.simpleName)>' to pkl eval"
+        )
+      else
+        if (value is String)
+          classHandlers[clazz].apply(value) ?? CoerceFail(path, "Unable to coerce '\(value)' into type \(clazz.simpleName)")
+        else
+          value
+    else
+      if (clazz is Class)
+        applyToClass(clazz, path)
+      else
+        coerceInput(value, reflect.TypeAlias(clazz).referent.referent.reflectee, path)
```

Comment:
Two observations:
* `classHandlers` only `Class` keys, so I'd do the `is TypeAlias` check at the top
* `value is String` fails for `value == null`
```suggestion
  if (clazz is TypeAlias) coerceInput(value, reflect.TypeAlias(clazz).referent.referent.reflectee, path)
  else if (!classHandlers.containsKey(clazz)) applyToClass(clazz, path)
  else if (value is String)
    classHandlers[clazz].apply(value) ?? CoerceFail(path, "Unable to coerce '\(value)' into type \(clazz.simpleName)")
  else
    value ?? ReadFail(
      path,
      "Not provided. Make sure you pass the argument '-p \(path.join(pathSeperator))=<\(clazz.simpleName)>' to pkl eval"
    )
```